core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1054,7 +1076,21 @@ class ReplicaManager(val config: KafkaConfig,
           startHighWaterMarksCheckPointThread()
           hwThreadInitialized = true
         }
+
+        // Add future replica to partition's map
+        val partitionsToAlterDirWithLogEndOffset = (partitionsBecomeLeader ++ partitionsBecomeFollower).flatMap { partition =>
+          logManager.getLog(partition.topicPartition, isFuture = true)
+        }.map { log =>
+          log.topicPartition -> BrokerAndInitialOffset(BrokerEndPoint(config.brokerId, "localhost", -1), log.logEndOffset)
+        }.toMap
+        partitionsToAlterDirWithLogEndOffset.keys.foreach(tp => getPartition(tp).get.getOrCreateReplica(Request.FutureLocalReplicaId))
+
+        // pause cleaning for partitions that are being moved and start ReplicaAlterDirThread to move replica from source dir to destination dir
+        partitionsToAlterDirWithLogEndOffset.keys.foreach(logManager.abortAndPauseCleaning)
+        replicaAlterDirManager.addFetcherForPartitions(partitionsToAlterDirWithLogEndOffset)
```

Comment:
Hmm, I am not sure that we should do line 1080-1090 on every LeaderAndIsr request especially the controller could re-send the same LeaderAndIsr request. It seems that it's enough to just do this on the very first LeaderAndIsr request (i.e., inside "if (!hwThreadInitialized) {").

Reply:
My concern with doing it only on the first LeaderAndIsrRequest is that this imposes two restrictions that it actually enforced now but may not be true in the future.

One restriction is that controller needs to send all partitions in the first LeaderAndIsrRequest to a broker. It is true as of now. But from design perspective broker should also be able to setup partition state properly if the partition is specified in the subsequent LeaderAndIsrRequest.

Another restriction is that broker always need to shutdown after receiving StopReplicaRequest with delete = false. Otherwise, when broker receives LeaderAndIsrRequest to be leader/follower for a partition, state of this partition may not be setup properly. This is also true for now. But it is just not very nice because ideally we should be able to send StopReplicaRequest and LeaderAndIsrRequest to stop/start a partition in a broker.

I have updated the patch to address the problem by only doing line 1080 - 1090 for partitions that are online and not already created before the broker receives this LeaderAndIsrRequest.